cpp/src/arrow/compute/exec/asof_join_benchmark.cc

```diff
@@ -0,0 +1,1023 @@
+#include "arrow/testing/future_util.h"
+#include "arrow/testing/generator.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/type.h"
+
+namespace arrow {
+namespace compute {
+
+static const char* time_col = "time";
+static const char* key_col = "id";
+static bool createdBenchmarkFiles = false;
+
+// requires export PYTHONPATH=/path/to/bamboo-streaming
+// calls generate_benchmark_files to create tables (feather files) varying in frequency,
+// width, key density for benchmarks. places generated files in benchmark/data. This
+// operation runs once at the beginning of benchmarking.
+static void DoSetup() {
+  if (!createdBenchmarkFiles) {
+    system(
+        "mkdir benchmark_data/ && python3 "
+        "~/summer/bamboo-streaming/bamboo/generate_benchmark_files.py");
+    createdBenchmarkFiles = true;
+  }
+}
+
+static void DoTeardown() { system("rm -rf benchmark_data/"); }
+
+static std::vector<std::string> generateRightHandTables(std::string freq, int width_index,
+                                                        int num_tables,
+                                                        int num_ids_index) {
+  auto const generate_file_name = [](std::string freq, std::string is_wide,
+                                     std::string num_ids, std::string num) {
+    return freq + "_" + is_wide + "_" + num_ids + num + ".feather";
+  };
+
+  std::string width_table[] = {"1_cols",  "10_cols", "20_cols",
+                               "40_cols", "80_cols", "100_cols"};   // 0 - 5
+  std::string num_ids_table[] = {"100_ids", "2000_ids", "5k_ids"};  // 0 - 2
+
+  std::string wide_string = width_table[width_index];
+  std::string ids = num_ids_table[num_ids_index];
+
+  std::vector<std::string> right_hand_tables;
+  for (int j = 1; j <= num_tables; j++) {
+    right_hand_tables.push_back(
+        generate_file_name(freq, wide_string, ids, std::to_string(j)));
+  }
+  return right_hand_tables;
+}
+
+// Wrapper to enable the use of RecordBatchFileReaders as RecordBatchReaders
+class RecordBatchFileReaderWrapper : public arrow::ipc::RecordBatchReader {
+  std::shared_ptr<arrow::ipc::RecordBatchFileReader> _reader;
+  int _next;
+
+ public:
+  virtual ~RecordBatchFileReaderWrapper() {}
+  explicit RecordBatchFileReaderWrapper(
+      std::shared_ptr<arrow::ipc::RecordBatchFileReader> reader)
+      : _reader(reader), _next(0) {}
+
+  virtual arrow::Status ReadNext(std::shared_ptr<arrow::RecordBatch>* batch) {
+    // cout << "ReadNext _next=" << _next << "\n";
+    if (_next < _reader->num_record_batches()) {
+      ARROW_ASSIGN_OR_RAISE(*batch, _reader->ReadRecordBatch(_next++));
+      // cout << "\t --> " << (*batch)->num_rows() << "\n";
+    } else {
+      batch->reset();
+      // cout << "\t --> EOF\n";
+    }
+
+    return arrow::Status::OK();
+  }
+
+  virtual std::shared_ptr<arrow::Schema> schema() const { return _reader->schema(); }
+};
+
+static std::tuple<arrow::compute::ExecNode*, int64_t, int, size_t>
```

Comment:
I'm not a big fan of using tuple as a return type here as it is rather opaque what these ints are representing.  Could you just create a simple struct?